lib/publiccloud/instance.pm

```diff
@@ -450,6 +452,13 @@ sub wait_for_ssh {
             sleep $delay;
         }    # end loop
 
+        if ($args{scan_ssh_host_key}) {
+            record_info('RESCAN', 'Rescanning SSH host key');
+            # Install server's ssh publicckeys to prevent authentication interactions
+            # or instance address changes during VM reboots.
+            script_run("ssh-keyscan $args{public_ip} | tee ~/.ssh/known_hosts /home/$testapi::username/.ssh/known_hosts");
```

Comment:
Why scan moved here at end ? It is a drastical logic change: why it is needed?

Reply:
When we do `softreboot` we run `wait_for_ssh`:
 * While checking if the system went down it still has the old address and old ssh host key.
 * While checking if the system went up it might have new IP address and we need to rescan the ssh host key.

Also when we check if the system already booted (`ssh xyz systemctl is-system-running`) we don't want to be bothered by the SSH key nor by the master socket so we ignore those.

We can speak about it.

Reply:
Please, consider that all code-body INSIDE this loop is NEVER executed when softreboot doing `shutdown`, it is skipped by `isok($exit_code) = false`: therefore this rescan will `not be applied`.

Rescan make sense only when booting up, at beginning before `nc` test: isn't it?

Reply:
No. Rescan does not work before neither after. It needs the SSH server to be ready.
That's why with my change we do:
1) `nc`
2) `ssh` without host key validation
3) rescan

Reply:
well, I trust you.
VRs welcome